nbs/20_comparison_others/py/05_02-gtex_whole_blood-sample_all_gene_pairs-mic.py

```diff
@@ -0,0 +1,271 @@
+
+from clustermatch import conf
+from clustermatch.utils import chunker
+
+# %% [markdown] tags=[]
+# # Settings
+
+# %% tags=[]
+DATASET_CONFIG = conf.GTEX
+GTEX_TISSUE = "whole_blood"
+GENE_SEL_STRATEGY = "var_pc_log2"
+
+METHOD_NAME = "mic"
+
+# %% tags=["parameters"]
+# this cell has the "parameters" tag
+
+# size of gene pair groups to process in parallel
+CHUNK_SIZE = 100
+
+# %% [markdown] tags=[]
+# # Paths
+
+# %% tags=[]
+INPUT_GENE_EXPR_FILE = (
+    DATASET_CONFIG["GENE_SELECTION_DIR"]
+    / f"gtex_v8_data_{GTEX_TISSUE}-{GENE_SEL_STRATEGY}.pkl"
+)
+display(INPUT_GENE_EXPR_FILE)
+
+assert INPUT_GENE_EXPR_FILE.exists()
+
+# %% tags=[]
+GENE_PAIRS_FILE_SUFFIX = (
+    DATASET_CONFIG["GENE_SELECTION_DIR"]
+    / f"gtex_v8_data_{GTEX_TISSUE}-{GENE_SEL_STRATEGY}.pkl"
+)
+display(GENE_PAIRS_FILE_SUFFIX)
+
+assert GENE_PAIRS_FILE_SUFFIX.exists()
+
+# %% tags=[]
+INPUT_DIR = GENE_PAIRS_FILE_SUFFIX.parent / "samples"
+display(INPUT_DIR)
+
+# %% tags=[]
+INPUT_GENE_PAIRS_FILE = INPUT_DIR / (
+    f"{GENE_PAIRS_FILE_SUFFIX.stem}-all-gene_pairs-sample_" + "{sample_id}" + ".pkl"
+)
+display(INPUT_GENE_PAIRS_FILE)
+
+INPUT_GENE_PAIRS_FILE_TEMPLATE = str(INPUT_GENE_PAIRS_FILE)
+display(INPUT_GENE_PAIRS_FILE_TEMPLATE)
+
+# %% tags=[]
+OUTPUT_DIR = DATASET_CONFIG["RESULTS_DIR"] / "comparison_others" / METHOD_NAME
+OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
+display(OUTPUT_DIR)
+
+# %% tags=[]
+OUTPUT_FILE_TEMPLATE = str(
+    OUTPUT_DIR / (INPUT_GENE_PAIRS_FILE.name[:-4] + f"-{METHOD_NAME}.pkl")
+)
+
+display(OUTPUT_FILE_TEMPLATE)
+
+# %% [markdown] tags=[]
+# # Data
+
+# %% [markdown] tags=[]
+# ## Gene expression
+
+# %% tags=[]
+gene_expr_dict = pd.read_pickle(INPUT_GENE_EXPR_FILE).T.to_dict(orient="series")
+
+# %% tags=[]
+len(gene_expr_dict)
+
+# %% tags=[]
+gene_expr_dict[list(gene_expr_dict.keys())[0]]
+
+# %% [markdown] tags=[]
+# ## Get all sample files
+
+# %% tags=[]
+all_sample_files = []
+
+sample_id = 0
+sample_file = Path(INPUT_GENE_PAIRS_FILE_TEMPLATE.format(sample_id=sample_id))
+display(sample_file)
+
+while sample_file.exists():
+    all_sample_files.append((sample_id, sample_file))
+
+    sample_id += 1
+    sample_file = Path(INPUT_GENE_PAIRS_FILE_TEMPLATE.format(sample_id=sample_id))
+
+# %% tags=[]
+display(len(all_sample_files))
+assert len(all_sample_files) > 0
+
+# %% tags=[]
+all_sample_files[:3]
+
+
+# %% [markdown] tags=[]
+# # Compute Maximal Information Coefficient (MIC)
+
+# %% [markdown] tags=[]
+# ## Functions
+
+# %% tags=[]
+def _mic(x, y):
+    """
+    Given two arrays (x and y), it computes MIC with the default parameters.
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=DeprecationWarning)
+
+        mine = MINE(alpha=0.6, c=15, est="mic_approx")
+        mine.compute_score(x, y)
+        return mine.mic()
+
+
+# %% tags=[]
+# testing
+_mic_val = _mic(np.random.rand(10), np.random.rand(10))
+display(_mic_val)
+assert _mic_val > 0.0
```

Comment:
I don't know anything about MIC, but is there a scenario, however unlikely, where _mic_val is equal to 0?

Reply:
Good point. MIC is defined in [0, 1], so technically this is possible. However, here I'm not testing whether MIC is working or not. This kind of `assert`s are just a translation of my visual assessment of the data/results to python code, so if I have to run it again I don't need to visually assess everything again just in case.